### PR TITLE
QPE Complete

### PR DESCRIPTION
**Pull request by Yelissal**

### CODE REVIEW

File: games/applications/run_qpe.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
+
+import os
+import pytest
+from qiskit import QuantumCircuit, QuantumRegister, ClassicalRegister
+#from applications import backends, run_qiskit_circuit
```
Comment (mtreinish) [inline games/applications/run_qpe.py:5]:
```suggestion
from applications import backends, run_qiskit_circuit
```

File: games/applications/run_qpe.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        qc.h(j)
+
+#Construct the phase gates and include matching gate representation as readme circuit
+def CPhase(angle, exponent):
+
+    qc = QuantumCircuit(1, name=f"U^{exponent}")
+    qc.p(angle*exponent, 0)
+    phase_gate = qc.to_gate().control(1)
+
+    return phase_gate, qc
```
Comment (mtreinish) [inline games/applications/run_qpe.py:46]:
```suggestion
```

File: games/applications/run_qpe.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        
+    #next apply the controlled unitary operation. this applies a phase to only the |1> state. 
+    repetition = 1
+    
+    #Phase of QFT
+    Phase = 2*math.pi*angle
+    
+    for j in (range(Cqubits)):
+        cp,Pgate= CPhase(Phase, repetition)
+        qc.append(cp, [j, Cqubits])
```
Comment (mtreinish) [inline games/applications/run_qpe.py:72]:
```suggestion
        qc.cp(Phase * repetition, j, Cqubits)
```

File: games/applications/run_qpe.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        repetition *= 2
+   
+    #define your Unitary gate 
+    Pgate,U = CPhase(Phase, repetition)
```
Comment (mtreinish) [inline games/applications/run_qpe.py:76]:
```suggestion
```

File: games/applications/run_qpe.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    #Lastly measure it 
+    qc.measure([qr[m] for m in range(Cqubits)], list(range(Cqubits)))
+
+    return qc
+
+#Benchmarking
+
+@pytest.mark.qiskit
+@pytest.mark.parametrize("optimization_level", [0, 1, 2, 3])
+#@pytest.mark.parametrize("backend", backends)
```
Comment (mtreinish) [inline games/applications/run_qpe.py:92]:
```suggestion
@pytest.mark.parametrize("backend", backends)
```

**Review by mtreinish — COMMENTED** [review]
This should bypass the failure you're hitting. The issue is there is a bug in qiskit's qasm generator when you have a custom controlled phase gate. It is generating overlapping definitions which can't be parsed (because the parser doesn't know which gate is which since they have duplicate names). You however don't need to do `PGate(...).control()` and can instead just use `qc.cp()` directly which will result in qasm that can be parsed.

**Review by danielleodigie — APPROVED** [review]
LGTM! Everything runs on my end :)

**Comment by mtreinish** [comment]
If this is ready to be merged can you mark this as ready to review and also you'll need to sign the CLA per: https://github.com/Qiskit/red-queen/pull/19#issuecomment-1184825909

File: games/applications/run_qpe.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
+# ------------------------------------------------------------------------------
+# Part of Red Queen Project.  This file is distributed under the Apache License.
+# See accompanying file /LICENSE for details.
+# ------------------------------------------------------------------------------
+
+# My code was inspired from the paper cited above. It provided a lot of directionality for the
+# structure of the code, however nothing is directly copied from it.
+# https://arxiv.org/pdf/2110.03137.pdf
```
Comment (keilydluna) [inline games/applications/run_qpe.py:8]: Code looks good, great Job! 

Consider linking the git repo from this paper in addition to the pdf.

File: games/applications/run_qpe.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+        if bit == 1:
+            num -= bit
+            binary += "1"
+        else:
+            binary += "0"
+        precision -= 1
+    return binary
+
+
+# The inverse QFT function was taken from the qiskit textbook from the section discussing
```
Comment (keilydluna) [inline games/applications/run_qpe.py:47]: Emphasize where you got this function at the beginning of your code as well. 

Refer to Caleb's grover's pull request for additional guidance.

**Review by raynelfss — APPROVED** [review]
Looks good to me!
Your fidelities are interesting as they don't exactly match the ones from the paper. However, this is potentially due to the noise simulated by the different backends, because running this on the AER Qasm simulator reutrns the expected status with a probability of 1.000.

File: games/applications/run_qpe.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+    benchmark.algorithm = f"Optimization level: {optimization_level} on {backend.name()}"
+    run_qiskit_circuit(benchmark, circ, backend, optimization_level, shots, expected_counts)
+
+
+if __name__ == "__main__":
+    quantum_phase_estimation(NUMQUBITS, SECRET_ANGLE).qasm(
+        filename=os.path.join(QASM_DIR, "qpe.qasm")
+    )
+
+    # Quantum_Phase_Estimation(i, SECRET_ANGLE).qasm(filename=os.path.join(QASM_DIR, f"qpe{i}.qasm"))
```
Comment (mtreinish) [inline games/applications/run_qpe.py:112]:
```suggestion
```

**Review by mtreinish — COMMENTED** [review]
The code LGTM, but before we can merge can you move the files to the new package directory `red_queen/games/applications/`

**Pull request closed without merging**